### ISSUE TITLE
i2c waveform for MPU9250 ESP32 interface

### ISSUE THREAD

**Issue by shomedas** [kriswiner/MPU-9250#397]
Hardware setup:
 MPU9250 Breakout --------- ESP32
 VDD ---------------------- 3.3V
 SDA ----------------------- 21
 SCL ----------------------- 22
 GND ---------------------- GND
NCS ----------------------- 3.3V
FSYNC -------------------- GND
AD0 ------------------------ GND

SDAL, SCL pull up to 3.3V is 4.5K

I2C bus speed is set 400KHz.
SMPLRT_DIV is set at 0x01 so that the sensor sampling frequency is at 500Hz. 

I captured the SDA & SCL waveforms which are as below. They dont look right to me, they look noisy. 

What do you say kris? Do you have any such wave forms captured for your 9250 setup on the i2c  of any microcontroller?

SCL stays high for 2ms and then there is square wave with 400KhZ frequency. This is repeated as seen from the following two images. Should not the SCL be  just a square wave with 400KhZ frequency

SCL signal:
![SCL_zoomed_out](https://user-images.githubusercontent.com/14357750/67591884-101f4180-f77c-11e9-8f2b-255518d73474.jpg)

SCL zoomed in:
![SCL_zoomed_in](https://user-images.githubusercontent.com/14357750/67591896-19101300-f77c-11e9-8f0d-eae822189987.jpg)

SDA is a square wave of frequency 500Hz(2ms time period). However it has a noisy, jittery region as shown below in the two Images

SDA signal:
![SDA_zoomed_out](https://user-images.githubusercontent.com/14357750/67591914-262d0200-f77c-11e9-8861-5cdfc18b7050.jpg)

SDA zoomed in
![SDA_zoomed_in](https://user-images.githubusercontent.com/14357750/67591922-2d541000-f77c-11e9-80e7-8ac769dcbb1d.jpg)

**Comment by kriswiner** [comment]
I2C wave form looks pretty good to me.

Did you verify that the interrupt is triggering at 500 Hz?

**Comment by shomedas** [comment]
i checked SDA waveform which had periodicity of 2ms ie the frequency was 500hz. is there any other way to check it.

Why is the SCL line high for significant time(2ms), then the 400Khz square wave(2.5us periodicity) occurs only for a short period. Isnt it supposed to be a just a square wave with periodicity of 2.5us?

**Comment by kriswiner** [comment]
Check the interrupt pin.

**Comment by shomedas** [comment]
Interrupt pin waveform. periodicity is 2ms. Looks ok i guess.

![interrupt_pin](https://user-images.githubusercontent.com/14357750/67597670-5a5aef80-f789-11e9-8a39-63b8ec4a7f40.jpg)

**Comment by shomedas** [comment]
However when I time through the code,  i get 63.5 ms between capture of two interuupts
My loop to check interrupt timing is

```
void loop()
{  
    if(newData == true) {  // On interrupt, read data

     Now = micros();
     newData = false;  // reset newData flag
     readMPU9250Data(MPU9250Data); // INT cleared on any read

    deltat = ((Now - lastUpdate));
    Serial.print("time between consecutive data packets in microsec= "); Serial.println(deltat);
    lastUpdate = Now;          
    }
}



```

**Comment by kriswiner** [comment]
Interrupt looks like it is firing at 1 kHz, is this right?

**Comment by shomedas** [comment]
No. its 500Hz. between two interrupts is 2ms. I checked on the CRO. The image is over exposed

**Comment by shomedas** [comment]
I don't understand why the timing shows 63.5 msec

**Comment by kriswiner** [comment]
OK, so you set the divide to 1 so you are getting 1 kHz/2 output data rate,
and the interrupt is firing t 500 Hz as it should?

Then, you are actually seeing 63 ms instead of 2 ms between data ready in
your loop?

**Comment by kriswiner** [comment]
Me either, it should show 2 ms, plus maybe a bit for I2C transactions...
What is the bandwidth? Try using very high bandwidth or no bandwidth

**Comment by shomedas** [comment]
which bandwidth are you referring to? The low pass filter bandwidth?

**Comment by kriswiner** [comment]
Yes, turn it off. or set it to 500 Hz.

**Comment by shomedas** [comment]
DLPF_CFG set to 0x07:  3600 Hz gyro bandwidth
A_DLPF_CFG set to 0x07:  460 Hz accel bandwidth

still the loop timing reads 63.5ms.

**Comment by shomedas** [comment]
I reduced the print statements length from 50 characters to 10 characters and timing changed from 63.5 ms to 17.5ms. 

Is it possible that some interrrupt cycles are lost if we print text in the loop.

**Comment by kriswiner** [comment]
UART is slow. Move the print statement out of the timing loop.

**Comment by shomedas** [comment]
The printing has to be taken out of the void loop().

Anyway this means that if we print anything in the void loop, we will loose some interrupt cycles.
